dbms/src/Functions/FunctionsLogical.h

```diff
@@ -377,154 +504,421 @@ class FunctionAnyArityLogical : public IFunction
+                }
+                else
+                {
+                    for (size_t i = 0; i < rows; ++i)
+                    {
+                        vec_res[i] = Impl::apply(const_val, col_data[i]);
+                    }
+                }
+            }
+            else
+            {
+                const auto & col_data = nullable_uint8_columns[0]->getData();
+                const auto & null_data = *null_maps[0];
+                if (const_val_is_null)
+                {
+                    for (size_t i = 0; i < rows; ++i)
+                        std::tie(vec_res[i], vec_res_is_null[i])
+                            = Impl::applyTwoNullable(const_val, const_val_is_null, col_data[i], null_data[i]);
+                }
+                else
+                {
+                    for (size_t i = 0; i < rows; ++i)
+                        std::tie(vec_res[i], vec_res_is_null[i])
+                            = Impl::applyOneNullable(col_data[i], null_data[i], const_val);
+                }
             }
         }
         else
         {
-            vec_res.resize(rows);
-            if constexpr (special_impl_for_nulls)
+            if (not_null_uint8_columns.size() == 2)
+            {
+                // case 1: 2 not null
+                AssociativeOperationImpl<Impl, 2>::execute(not_null_uint8_columns, vec_res);
+            }
+            else if (not_null_uint8_columns.size() == 1 && nullable_uint8_columns.size() == 1)
+            {
+                // case 2: 1 not null + 1 nullable
+                const auto & col_1 = not_null_uint8_columns[0]->getData();
+                const auto & col_2 = nullable_uint8_columns[0]->getData();
+                const auto & null_map_2 = *null_maps[0];
+                for (size_t i = 0; i < rows; ++i)
+                {
+                    std::tie(vec_res[i], vec_res_is_null[i])
+                        = Impl::applyOneNullable(col_2[i], null_map_2[i], col_1[i]);
+                }
+            }
+            else
             {
-                vec_input_has_null.assign(rows, (UInt8)0);
-                vec_res_not_null.assign(rows, (UInt8)0);
+                // case 3: 2 nullable
+                NullableAssociativeOperationImpl<Impl, 2>::execute(
+                    nullable_uint8_columns,
+                    null_maps,
+                    vec_res,
+                    vec_res_is_null);
             }
         }
+        if (result_is_nullable)
+        {
+            block.getByPosition(result).column = ColumnNullable::create(std::move(col_res), std::move(col_res_is_null));
+        }
+        else
+        {
+            block.getByPosition(result).column = std::move(col_res);
+        }
+    }
 
-        /// Convert all columns to UInt8
-        UInt8ColumnPtrs uint8_in;
-        Columns converted_columns;
-
-        for (size_t index = 0; index < in.size(); index++)
+    void handleMultipleInputColumns(
```

Comment:
Better add one full explanation for all the parameters, and describe the differences for other two handleXXXX methods if any.

Reply:
Done